src/passes/pass-utils.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_passes_pass_utils_h
+#define wasm_passes_pass_utils_h
+
+#include <unordered_set>
+
+#include <pass.h>
+#include <wasm.h>
+
+namespace wasm {
+
+namespace PassUtils {
+
+using FuncSet = std::unordered_set<Function*>;
+
+// A wrapper around a parallel pass that filters it to run run only on select
+// functions.
+struct FilteredPass : public Pass {
+  std::unique_ptr<Pass> create() override {
+    // Function-parallel passes get a new instance per function. Create a copy
+    // of the wrapped pass along with ourselves.
+    return std::make_unique<FilteredPass>(pass->create(), relevantFuncs);
+  }
+
+  FilteredPass(std::unique_ptr<Pass> pass, const FuncSet& relevantFuncs)
+    : pass(std::move(pass)), relevantFuncs(relevantFuncs) {}
+
+  bool isFunctionParallel() override {
+    assert(pass->isFunctionParallel());
+    return true;
+  }
+
+  void runOnFunction(Module* module, Function* func) override {
+    if (!relevantFuncs.count(func)) {
+      return;
+    }
+
+    // The pass runner calling us set our pass runner, which we must do for the
+    // wrapped pass.
+    pass->setPassRunner(getPassRunner());
+    pass->runOnFunction(module, func);
+  }
+
+  bool modifiesBinaryenIR() override { return pass->modifiesBinaryenIR(); }
+
+  bool invalidatesDWARF() override { return pass->invalidatesDWARF(); }
+
+  bool addsEffects() override { return pass->addsEffects(); }
+
+  bool requiresNonNullableLocalFixups() override {
+    return pass->requiresNonNullableLocalFixups();
+  }
+
+private:
+  std::unique_ptr<Pass> pass;
```

Comment:
Instead of having an indirection here, it might make sense to use templates to store the pass inline in the `FilteredPass<P>` (or even as the supertype of `FilteredPass<P>`, which would avoid having to implemented `modifiesBinaryenIR`, etc.). WDYT?

Reply:
I don't see how `FilteredPass` can be templated over the pass. The pass arrives from `FilteredPassRunner::doAdd` which does not have that information - it just gets a pass instance. (Templating `doAdd` might be possible but that would be a large change I think and I'm not sure if it can work or not.)

Reply:
Ah right, if you don't statically know the full type at every callsite, templates won't work. Makes sense.